collect_app/src/main/java/org/odk/collect/android/formmanagement/FormMetadataParser.kt

```diff
@@ -91,17 +89,15 @@ object FormMetadataParser {
      * Returns the XPath path for the first geopoint in the primary instance that is before the given
      * reference and not in a repeat.
      */
-    private fun getInstanceGeoPointBeforeXPath(firstTopLevelBodyGeoPoint: String?, model: Element, mainInstance: Element): String? {
+    private fun getInstanceGeoPointBeforeXPath(model: Element, mainInstance: Element): String? {
         val mainInstanceRoot = mainInstance.getElement(0)
         for (elementId in 0 until mainInstanceRoot.childCount) {
             val child = mainInstanceRoot.getElement(elementId)
             val ref = "/${mainInstanceRoot.name}/${child.name}"
-            if (ref == firstTopLevelBodyGeoPoint) {
```

Comment:
> The goal of this function is to only return the target of a setvalue action if it comes before a user-facing geopoint.

Hmm `setgeopoint` action represents [Geolocation at survey start](https://docs.getodk.org/form-question-types/#geolocation-at-survey-start) right? It's the action triggered after opening a form so how could anything go before? Isn't it always the first geo element?

Reply:
Currently we only exposed the `setvalue` action triggered by the new instance event in XSLForm. You can also have `setvalue` triggered by other events like value change. Neither is guaranteed to be in any specific order. In the case of a `setvalue` triggered by value change, the action would be in the body.

Reply:
I think we need to take a different approach on this one. When we had the processed form definition from JR we could do smarter things but I don’t think they’re practical in this context. Maybe we could do something like 
- iterate through the binds, build a set of  the refs that have type geopoint
- iterate through the primary instance, return the first ref that’s in the set of geopoints

That makes some assumptions that are not guaranteed by XForms, most notably that refs in the body are in the same order in the model. But in practice they usually are so I think it’s reasonable.

Reply:
Alright, we can do that. However, my question is: if the order isn't guaranteed, why check the primary instance rather than just the binds? I assume the order of the binds is even less predictable?

Reply:
Yes, exactly.

Reply:
Ok so please take a look at the new implementation.